File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/VariableUnused.cs
﻿using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Tests.Diagnostics
{
    public class VariableUnused
    {
        void F1()
        {
            var packageA = DoSomething("Foo", "1.0");
            var packageB = DoSomething("Qux", "1.0");

            var localRepository = new Cl { packageA, packageB }; // Noncompliant
//              ^^^^^^^^^^^^^^^

            using (var x = new StreamReader("")) // Compliant
            {
                var v = 5; // Noncompliant {{Remove the unused local variable 'v'.}}
            }

            int a; // Noncompliant
            var b = (Action<int>)(
                _ =>
                {
                    int i; // Noncompliant
                    int j = 42;
                    Console.WriteLine("Hello, world!" + j);
                });

            b(5);

            string c;
            c = "Hello, world!";
            Console.WriteLine(c);

            var d = "";
            var e = new List<String> { d };
            Console.WriteLine(e);
        }

        private object DoSomething(string foo, string p1)
        {
            throw new NotImplementedException();
        }

        void F2(int a)
        {
        }
    }

    internal class Cl : List<object>
    {
    }

    public class Lambdas
    {
        public void Foo(List<int> list)
        {
            list.Select(item => 1);

            list.Select(item =>
                {
                    var value = 1; // Noncompliant
                    return item;
                });
        }
    }

    public class WithLocalFunctions
    {
        public void Method()
        {
            void Foo(List<int> list)
            {
                list.Select(item => 1);

                list.Select(item =>
                {
                    var value = 1; // Noncompliant
                    return item;
                });
            }

            static void Bar(List<int> list)
            {
                list.Select(item => 1);

                list.Select(item =>
                {
                    var value = 1; // Noncompliant
                    return item;
                });
            }
        }
    }

    //https://github.com/SonarSource/sonar-dotnet/issues/3137
    public class Repro_3137
    {
        public void GoGoGo(Logger log)
        {
            using var _ = log.BeginScope("XXX"); // Noncompliant FP, discard pattern cannot be used

            // This is a real noncompliant that should be reported. File.Create(path).Dispose(); should be used instead
            using var stream = File.Create("path"); // Noncompliant
        }

        public class Logger
        {
            public IDisposable BeginScope(string scope)
            {
                return null;
            }
        }
    }
}

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/VariableUnused.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq

Namespace Tests.Diagnostics
    Public Class VariableUnused
        Private Sub F1()
            Dim packageA = DoSomething("Foo", "1.0")
            Dim packageB = DoSomething("Qux", "1.0")
            Dim localRepository = New Cl From { packageA, packageB }
'               ^^^^^^^^^^^^^^^

            Using x = New StreamReader("")
                Dim v = 5 ' Noncompliant {{Remove the unused local variable 'v'.}}
            End Using

            Dim a As Integer ' Noncompliant
            Dim b = CType((Sub(__)
                               Dim i As Integer ' Noncompliant
                               Dim j As Integer = 42
                               Console.WriteLine("Hello, world!" & j)
                           End Sub), Action(Of Integer))
            b(5)
            Dim c As String
            c = "Hello, world!"
            Console.WriteLine(c)
            Dim d = ""
            Dim e = New List(Of String) From {
                d
            }
            Console.WriteLine(e)
        End Sub

        Private Function DoSomething(ByVal foo As String, ByVal p1 As String) As Object
            Throw New NotImplementedException()
        End Function

        Private Sub F2(ByVal a As Integer)
        End Sub
    End Class

    Friend Class Cl
        Inherits List(Of Object)
    End Class

    Public Class Lambdas
        Public Sub Foo(ByVal list As List(Of Integer))
            list.[Select](Function(item) 1)
            list.[Select](Function(item)
                              Dim value = 1 ' Noncompliant
                              Return item
                          End Function)
        End Sub
    End Class
End Namespace
